DOC: fix remaining doc files for refguide_check

pdebuyl opened the pull request:
Related to #14970 

Fix remaining files for refguide_check.

```
doc/source/f2py/getting-started.rst
doc/source/reference/arrays.nditer.rst
doc/source/user/c-info.python-as-glue.rst
```

Notes:
- There remained some formatting issues that I put in separate commits.
- Most problems stemmed from cython or f2py modules that cannot be compiled on the fly with the current build/test system. Those are skipped, see https://github.com/numpy/numpy/issues/14970#issuecomment-595859279
- There was a separate discrepancy in arrays.nditer.rst where some examples made incorrect use of the "external_loop" flag.
- In arrays.nditer.rst, the loop output was unintuitive in some places and was fixed by assigning the loop output to a dummy variable, as discussed here: https://github.com/numpy/numpy/pull/15519#discussion_r375909343

With this PR, `python tools/refguide_check.py --rst` outputs

```
OK: all checks passed!
```

and the sphinx build proceeds without warning (apart from a minor data clipping warning for imshow).

After merging, the NumPy project should consider testing via pytest and suitable plugins, including [astropy's pytest-doctestplus](https://github.com/astropy/pytest-doctestplus/), see https://github.com/numpy/numpy/pull/15703#discussion_r388093856 and migrating the doc accordingly.

I test pytest on the files I edited and most errors stemmed from whitespace issues or specific usages (matplotlib functions return a repr for the figure object, those are not included in the current docs).

pdebuyl committed:
DOC: skip f2py test for refguide_check

pdebuyl committed:
DOC: fix formatting of output and traceback

pdebuyl committed:
DOC: remove external_loop flags to nditer

The external_loop prevent the iterations to go over items, which is not
the intended effect in the document.

pdebuyl committed:
DOC: skip cython-dependent test

This is likely a temporary measure to enable continuous integration
testing of the docs.

pdebuyl committed:
DOC: skip f2py-dependent test

This is likely a temporary measure to enable continuous integration
testing of the docs.

pdebuyl committed:
DOC: fix formatting of doc tests

Fix whitespace issues.

Fix formatting for current NumPy version.

mattip commented:
Rather than skipping the f2py tests we could theoretically try to execute them:
```
.. for doctests

  >>> import numpy.f2py, os
  >>> with open('fib1.f') as fid:
  ...     source = fid.read()
  >>> numpy.f2py.compile(source, modulename='fib1')
```

but this does not work. The doctests are executed in a temporary directory, and have no access to the `fib1.f` file.

pdebuyl commented:
I noticed as well.

Running `pytest --doctest-glob '*.rst'` executes the tests with the path when executing the command as the current working directory for the tests. But this requires replacing refguide_check.py with pytest.

mattip commented:
> But this requires replacing refguide_check.py with pytest.

I am OK with skipping all the f2py compilation doctests for now then.

mattip commented on doc/source/reference/arrays.nditer.rst:
```diff
@@ -28,7 +28,7 @@ using the standard Python iterator interface.
     ...     print(x, end=' ')
     ...
-    0 1 2 3 4 5
+    0 1 2 3 4 5 
```
extra space at the end of the line

mattip commented:
Other than some extra spaces at the end of lines, this looks good to me.

pdebuyl commented:
Hi @mattip thank you for the feedback. There might be an issue with the end-of-line marker on github. I did put the extra spaces. Did you check with a local copy?

mattip commented:
> I did put the extra spaces

We don't really like extra spaces on the end of lines and refguide-check does not require them, so it would be better to remove them.

pdebuyl commented:
I used pytest to check the spacing of all tests, and the end-of-line space was required here. If we do move to pytest ultimately, those will be required.

Let me know if I should remove them despite this perspective.

mattip commented:
Isn't there a way to skip all the f2py doctests using the SKIPLIST in refguide-check? Also - please remove the spaces at the end of the lines.

rgommers commented:
I agree, a bit of context:

> Isn't there a way to skip all the f2py doctests using the SKIPLIST in refguide-check? 

This (skiplist) is how it also works in SciPy, and it's nicer because `#doctest: SKIP` is visual noise in examples (which is what doctests are primarily).

> Also - please remove the spaces at the end of the lines.

Rationale: someone else's editor will strip these spaces again. We should just fix the printing behavior if needed (it doesn't really seem needed now).

pdebuyl committed:
Revert "DOC: skip f2py-dependent test"

This reverts commit 65f86ef65e8eab7dc640432545c8ee4ba29b2420.

pdebuyl committed:
Revert "DOC: fix formatting of doc tests"

This reverts commit be6cb634c368a09a855e94e0edab2e449d32cfbc.

pdebuyl committed:
Revert "DOC: skip f2py test for refguide_check"

This reverts commit 02c4e5c04579c849dbb8a3179612b534af85fa2b.

pdebuyl committed:
ignore f2py dependent rst files

pdebuyl commented:
changes done, I tested locally and wait for the CI checks.

Some SKIP remain in arrays.nditer.rst as a small part of that file only relies on a compiled extension. Some examples there also use timeit which is not valid outside of IPython/jupyter

pdebuyl commented:
@mattip I have addressed the review items.

mattip commented on tools/refguide_check.py:
```diff
@@ -119,6 +119,8 @@
     'doc/source/release',
     'c-info.ufunc-tutorial.rst',
+    'c-info.python-as-glue.rst',
+    'getting-started.rst',
```
I got scared for a second, until I realized that `gettting-started.rst` is not the general beginner guide, rather the one inside the `f2py` directory. Could you rename the file to `f2py.getting-started.rst` or something that would prevent a future conflict with a `getting-started.rst` file for another task?

pdebuyl replied:
Sure. Here, as a contributor I hesitated to rename a file.

mattip commented on doc/source/reference/arrays.nditer.rst:
```diff
@@ -736,30 +737,30 @@ as our native Python/NumPy code did.
     100 loops, best of 3: 11.8 ms per loop
 
-    >>> np.all(sum_squares_cy(a, axis=-1) == np.sum(a*a, axis=-1))
+    >>> np.all(sum_squares_cy(a, axis=-1) == np.sum(a*a, axis=-1)) # doctest: +SKIP
```
All these skips are annoying. Would it work to
- split the file in two, move all the cython code to `arrays.nditer.cython.rst`
- add `arrays.nditer.cython.rst` to the skiplist
- use an `.. include:: arrays.nditer.cython.rst` so that sphinx will render the documents as before
- add `.. for doctests` comments just before the include and at the top of the new file to explain why this was done

pdebuyl replied:
Tentative update pushed

pdebuyl committed:
DOC: rename f2py file getting-started.rst

pdebuyl committed:
DOC: Split file arrays.nditer.rst

Place the Cython-dependent section in separate file.

Ignore the new file from the doctests with an RST_SKIPLIST entry in
refguide_check.py

mattip commented:
Looks good now thanks. At some point we should refactor the redundant parts of [`arrays.nditer.rst`](https://numpy.org/devdocs/reference/arrays.nditer.html) and ['nditer'](https://numpy.org/devdocs/reference/generated/numpy.nditer.html) notes.

mattip commented:
Thanks @pdebuyl
